clippy_lints/src/float_literal.rs

```diff
@@ -126,13 +128,23 @@ impl<'tcx> LateLintPass<'tcx> for FloatLiteral {
                         },
                     );
                 }
-            } else if digits > max as usize && count_digits(&float_str) < count_digits(sym_str) {
+            } else if digits > max as usize && count_digits(&float_str) < digits {
+                if digits >= 40 && is_const_item(cx, expr) {
```

Comment:
This will not match on `const` items which are impl items or trait items. Also, `static` should probably be treated the same way.

```suggestion
                if digits >= 40 && matches!(expr_use_ctxt(cx, expr).use_node(cx), ExprUseNode::ConstStatic(_)) {
```

Tests should be added for all those cases as well.

Reply:
In the embedded world for example, you might want to ensure that your `f64` constant is present only once in read-only memory for example, especially if you pass them by reference. You would use `static` in some cases instead of counting on linker optimizations to unify them.